### PR TITLE
feat(cincinnati): adds download and image building functionality for Cincinnati graph data

### LINKED ISSUE #94: Download latest upgrade graph image
Download latest cincinnati graph data image when mirror:ocp:graph is true in the imageset-config and generate the require `Cincinnati` manifest

### PR DESCRIPTION
**Pull request by jpower432**
# Description

To interact with OSUS, a Cincinnati graph image must be available in the registry. This feature downloads the graph contents during imageset creation and builds the image during image publishing.

Fixes #94 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [X] Manually tested with m2d and m2m workflows
- [X] Unit tests added for manifest creation 

#Testing Scenarios
```
apiVersion: mirror.openshift.io/v1alpha2
kind: ImageSetConfiguration
mirror:
  ocp:
    graph: true
    channels:
    - name: stable-4.9
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.9
      headsOnly: false
      packages:
        - name: cincinnati-operator
          startingBundle: update-service-operator.v4.9.1
```
Install the latest 4.9 release, with the graph and install the Cincinnati operator. The release signature will be located in the results directory under the `release-signatures` along with the `updateService.yaml`

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by jpower432** [comment]
/hold

Hold until integration tests are completed  cc @jharmison-redhat.

**Comment by wking** [comment]
I'm just leaning in here without much context, but is something in here applying the signature ConfigMap manifest (mentioned [here][1] with `oc apply -f ${REMOVABLE_MEDIA_PATH}/mirror/config/<image_signature_file>`) to target clusters?

[1]: https://docs.openshift.com/container-platform/4.10/updating/updating-restricted-network-cluster.html#update-mirror-repository_updating-restricted-network-cluster

**Comment by jpower432** [comment]
> I'm just leaning in here without much context, but is something in here applying the signature ConfigMap manifest (mentioned [here](https://docs.openshift.com/container-platform/4.10/updating/updating-restricted-network-cluster.html#update-mirror-repository_updating-restricted-network-cluster) with `oc apply -f ${REMOVABLE_MEDIA_PATH}/mirror/config/<image_signature_file>`) to target clusters?

Hey @wking. We have another PR for the generation of the ConfigMap #357. `oc-mirror` does not do any actually manifest application, just generation.

File: pkg/archive/archive.go [hunk 1]
```diff
@@ -256,8 +269,16 @@ func blobInArchive(file string) string {
 }
 
 func includeFile(fpath string) bool {
+	includeFiles := map[string]struct{}{
+		config.InternalDir:         {},
+		"catalogs":                 {},
```
Comment (dinhxuanvu) [inline pkg/archive/archive.go:274]: Just a nit. This random string among other constants is really ugly to see. At some point, it should be a variable/constant somewhere so it won't be so lonely here.

File: pkg/cli/mirror/helm.go [hunk 2]
```diff
@@ -62,7 +62,7 @@ func (h *HelmOptions) PullCharts(ctx context.Context, cfg v1alpha2.ImageSetConfi
 	c := downloader.ChartDownloader{
 		Out:     os.Stdout,
 		Keyring: "",
-		Verify:  downloader.VerifyNever,
+		Verify:  downloader.VerifyIfPossible,
```
Comment (dinhxuanvu) [inline pkg/cli/mirror/helm.go:65]: Would like to know the reason behind verification mode change here.

File: pkg/cli/mirror/manifests.go [hunk 3]
```diff
@@ -210,6 +210,27 @@ func generateCatalogSource(name string, dest reference.DockerImageReference) ([]
 	return cs, nil
 }
 
+func generateUpdateService(name string, releaseRepo, graphDataImage reference.DockerImageReference) ([]byte, error) {
+
+	obj := map[string]interface{}{
```
Comment (dinhxuanvu) [inline pkg/cli/mirror/manifests.go:215]: Is there an object type somewhere for `UpdateService` in openshift library that we can use instead of having to manually construct it like this?

File: pkg/cli/mirror/manifests.go [hunk 4]
```diff
@@ -261,3 +282,16 @@ func WriteCatalogSource(mapping image.TypedImageMapping, dir string) error {
 	logrus.Infof("Wrote CatalogSource manifests to %s", dir)
 	return nil
 }
+
+// WriteCatalogSource will generate a CatalogSource object and write it to disk
```
Comment (dinhxuanvu) [inline pkg/cli/mirror/manifests.go:286]: Comment is not matching the func. Please doublecheck this.

**Review by dinhxuanvu — COMMENTED** [review]
Reviewed. A few comments.

File: pkg/archive/archive.go (hunk 1 above)
Comment (jpower432) [reply pkg/archive/archive.go:274]: Totally agree. It's set as a const in the `cli` package, but I will move it to `config` so it can be referenced like the others.

File: pkg/cli/mirror/manifests.go (hunk 3 above)
Comment (jpower432) [reply pkg/cli/mirror/manifests.go:215]: AFAIK there is not, but I will do some research to double-check this.

File: pkg/cli/mirror/helm.go (hunk 2 above)
Comment (jpower432) [reply pkg/cli/mirror/helm.go:65]: @dinhxuanvu Thanks for catching this. I meant to separate this out as it does not have to do with the feature. The motivation behind the mode change is to allow signed Helm chart to be verified. I will remove and put in a new PR.

File: pkg/cli/mirror/manifests.go (hunk 3 above)
Comment (jpower432) [reply pkg/cli/mirror/manifests.go]: Was able to incorporate this, but used the `UpdateServiceSpec` type in my own type so the manifest was not generated the status fields.

File: pkg/cli/mirror/manifests.go [hunk 5]
```diff
@@ -261,3 +299,16 @@ func WriteCatalogSource(mapping image.TypedImageMapping, dir string) error {
 	logrus.Infof("Wrote CatalogSource manifests to %s", dir)
 	return nil
 }
+
+// WriteUpdateService will generate an UpdateService object and write it to disk
+func WriteUpdateService(release, graph image.TypedImage, dir string) error {
+	updateService, err := generateUpdateService("update-service", release.Ref, graph.Ref)
```
Comment (dinhxuanvu) [inline pkg/cli/mirror/manifests.go:305]: Are we gonna have a problem with the fact we hardcode the name of the UpdateService here which essentially enforce "there will be only one UpdateService and will be overwritten in subsequent runs"?

File: pkg/cli/mirror/manifests.go [hunk 6]
```diff
@@ -261,3 +299,16 @@ func WriteCatalogSource(mapping image.TypedImageMapping, dir string) error {
...
+	if err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(filepath.Join(dir, "updateService.yaml"), updateService, os.ModePerm); err != nil {
```
Comment (dinhxuanvu) [inline pkg/cli/mirror/manifests.go:309]: Same principle from above question applies on this local hardcoded filename.

**Review by dinhxuanvu — COMMENTED** [review]
Looking good. Just a couple more questions.

File: pkg/cli/mirror/manifests.go (hunk 6 above)
Comment (jpower432) [reply pkg/cli/mirror/manifests.go:309]: This should not be an issue given that we write this into a new result directory every run. Only one `updateService.yaml` would ever be generated per run.

File: pkg/cli/mirror/manifests.go (hunk 5 above)
Comment (jpower432) [reply pkg/cli/mirror/manifests.go:305]: Good question. I think this is probably okay because AFAIK one instance of the `UpdateService` would be run in a cluster at any given time. Since we are resolving the graph image to digest, I would think we would want it overwritten. 

Maybe @wking can shed some light here. Is there a use case where overwriting the UpdateService manifest could cause a problem?

Comment (wking) [reply pkg/cli/mirror/manifests.go:305]: it is theoretically possible to run multiple UpdateService instances in the same cluster.  That seems pretty niche though.  As long as you have a name for your UpdateService that's fairly unique to `oc-mirror`, I expect conflicts with other tooling or manually-created UpdateService would be unlikely.  And we also currently have [rbhz#1939855][1], making it harder than it should be to get the update service to notice an UpdateService bump to `graphDataImage`.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1939855

Comment (jpower432) [reply pkg/cli/mirror/manifests.go:305]: @dinhxuanvu I agree. I am good with updating the name on the manifest to something like `update-service-oc-mirror`. 
Also thanks for the info @wking on bumping the `graphDataImage`, it is something we should keep in mind though since we do not actually apply any manifests to the cluster, it would not impact `oc-mirror` functionality. Possibly warrants a warning in docs.

**Comment by dinhxuanvu** [comment]
@jpower432 The sanity test doesn't look happy for some reason. Can you do a round of `go mod tidy/vendor` when you push your next patch of changes? Thanks.

**Review by dinhxuanvu — APPROVED** [review]
/lgtm

**Comment by jharmison-redhat** [comment]
Tested f542a3c using the [oc-mirror-e2e environment on commit c686c5a](https://github.com/jharmison-redhat/oc-mirror-e2e/tree/c686c5a7e24c8a7901bc9b2557d9b7e4b7496ced) and the following:

`scenario.yaml`:

```yaml
---
cluster_name: disco
openshift_version: "4.9"

scenario_stubs:
  mirror_method: to_registry
  registry_type: docker_registry

imageset_config_override: |-
  apiVersion: mirror.openshift.io/v1alpha2
  kind: ImageSetConfiguration
  mirror:
    platform:
      graph: true
      channels:
      - name: stable-4.9
    operators:
      - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.9
        full: true
        packages:
          - name: cincinnati-operator
            startingBundle: update-service-operator.v4.9.1

oc_mirror_source:
  repo: https://github.com/jpower432/bundle.git
  version: feat/issue-94
```

Commands:
- `make clean run ANSIBLE_PLAYBOOKS='create test'`
- `scp -3 <registry>:mirror-content/oc-mirror-workspace/results-1649114059/release-signatures/signature-sha256-2eafde815e543b92.json <bastion>:release-signature-cm.json` (adapt results dir and signature to time of mirror and hostnames to whatever you need to be able to SSH in the environment)
- `ssh <bastion>`
  - `export KUBECONFIG=~/install/auth/kubeconfig`
  - `oc create cm -n openshift-config --from-file=registry.internal.redhat4govaws.io=/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem --from-file=updateservice-registry=/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem registry-config`
  - `oc patch image.config.openshift.io cluster -p '{"spec":{"additionalTrustedCA":{"name":"registry-config"}}}' --type=merge`
  -
  ```
  oc apply -f- << EOF
  apiVersion: v1
  kind: Namespace
  metadata:
    name: openshift-update-service
    annotations:
      openshift.io/node-selector: ""
    labels:
      openshift.io/cluster-monitoring: "true"
  EOF
  ```
  -
  ```
  oc apply -n openshift-update-service -f- << EOF
  apiVersion: operators.coreos.com/v1
  kind: OperatorGroup
  metadata:
    name: update-service-operator-group
  spec:
    targetNamespaces:
    - openshift-update-service
  EOF
  ```
  -
  ```
  oc apply -n openshift-update-service -f- << EOF
  apiVersion: operators.coreos.com/v1alpha1
  kind: Subscription
  metadata:
    name: update-service-subscription
  spec:
    channel: v1
    installPlanApproval: "Automatic"
    source: "redhat-operator-index"
    sourceNamespace: "openshift-marketplace"
    name: "cincinnati-operator"
  EOF
  ```
  - `oc apply -f release-signature-cm.json` (scp'd from previous step)
  - `oc apply -f mirror-content/updateService.yaml -n openshift-update-service`

Validated Update Service launched correctly with the following:

```shell
$ oc get deploy -n openshift-update-service update-service-oc-mirror
NAME                       READY   UP-TO-DATE   AVAILABLE   AGE
update-service-oc-mirror   2/2     2            2           11m
```

```shell
$ oc logs -n openshift-update-service deploy/update-service-oc-mirror -c graph-builder | tail
Found 2 pods, using pod/update-service-oc-mirror-6444b94dbf-5wwwj
[2022-04-05T00:30:06Z INFO  cincinnati::plugins::internal::edge_add_remove] [4.9.25+amd64]: removing previous 4.8.27+amd64 by regex
[2022-04-05T00:30:06Z INFO  cincinnati::plugins::internal::edge_add_remove] [4.9.25+amd64]: removing previous 4.8.28+amd64 by regex
[2022-04-05T00:30:06Z INFO  cincinnati::plugins::internal::edge_add_remove] [4.9.25+amd64]: removing previous 4.8.29+amd64 by regex
[2022-04-05T00:30:06Z INFO  cincinnati::plugins::internal::edge_add_remove] [4.9.25+amd64]: removing previous 4.8.30+amd64 by regex
[2022-04-05T00:30:06Z INFO  cincinnati::plugins::internal::edge_add_remove] [4.9.25+amd64]: removing previous 4.8.31+amd64 by regex
[2022-04-05T00:30:06Z INFO  cincinnati::plugins::internal::edge_add_remove] [4.9.25+amd64]: removing previous 4.8.32+amd64 by regex
[2022-04-05T00:30:06Z INFO  cincinnati::plugins::internal::edge_add_remove] [4.9.25+amd64]: removing previous 4.8.33+amd64 by regex
[2022-04-05T00:30:06Z INFO  cincinnati::plugins::internal::edge_add_remove] [4.9.25+amd64]: removing previous 4.8.34+amd64 by regex
[2022-04-05T00:30:06Z INFO  cincinnati::plugins::internal::edge_add_remove] [4.9.25+amd64]: removing previous 4.8.35+amd64 by regex
[2022-04-05T00:30:06Z DEBUG graph_builder::graph] graph update completed, 43 valid releases
```

tl;dr:
/lgtm

**Comment by jpower432** [comment]
@dinhxuanvu Can you please give the last two commits a look before I can cancel the hold? Fixes to the feature per integration test feedback. Thanks!

**Review by dinhxuanvu — APPROVED** [review]
/lgtm
